engine/hatchery/vsphere/vsphere.go

```diff
@@ -2,311 +2,392 @@ package vsphere
+	NewVirtualMachine(ctx context.Context, cloneSpec *types.VirtualMachineCloneSpec, ref *types.ManagedObjectReference) (*object.VirtualMachine, error)
+	RenameVirtualMachine(ctx context.Context, vm *object.VirtualMachine, newName string) error
+	ReconfigureVirtualMachine(ctx context.Context, vm *object.VirtualMachine, config types.VirtualMachineConfigSpec) error
+	MarkVirtualMachineAsTemplate(ctx context.Context, vm *object.VirtualMachine) error
+	WaitForVirtualMachineShutdown(ctx context.Context, vm *object.VirtualMachine) error
+	WaitForVirtualMachineIP(ctx context.Context, vm *object.VirtualMachine) error
+	LoadFolder(ctx context.Context) (*object.Folder, error)
+	SetupEthernetCard(ctx context.Context, card *types.VirtualEthernetCard, ethernetCardName string, network object.NetworkReference) error
+	LoadNetwork(ctx context.Context, name string) (object.NetworkReference, error)
+	LoadResourcePool(ctx context.Context) (*object.ResourcePool, error)
+	LoadDatastore(ctx context.Context, name string) (*object.Datastore, error)
+	ProcessManager(ctx context.Context, vm *object.VirtualMachine) (*guest.ProcessManager, error)
+	StartProgramInGuest(ctx context.Context, procman *guest.ProcessManager, req *types.StartProgramInGuest) (int64, error)
+}
 
-// New instanciates a new Hatchery vsphere
-func New() *HatcheryVSphere {
-	s := new(HatcheryVSphere)
-	s.GoRoutines = sdk.NewGoRoutines()
-	s.Router = &api.Router{
-		Mux: mux.NewRouter(),
+func NewVSphereClient(vclient *govmomi.Client, datacenter string) VSphereClient {
+	return &vSphereClient{
+		vclient:        vclient,
+		requestTimeout: 15 * time.Second,
+		datacenter:     datacenter,
 	}
-	return s
 }
 
-var _ hatchery.InterfaceWithModels = new(HatcheryVSphere)
+type vSphereClient struct {
+	datacenter     string
+	vclient        *govmomi.Client
+	requestTimeout time.Duration
+}
+
+func (c *vSphereClient) finder(ctx context.Context) (*find.Finder, error) {
+	finder := find.NewFinder(c.vclient.Client, false)
 
-// Init cdsclient config.
-func (h *HatcheryVSphere) Init(config interface{}) (cdsclient.ServiceConfig, error) {
-	var cfg cdsclient.ServiceConfig
-	sConfig, ok := config.(HatcheryConfiguration)
-	if !ok {
-		return cfg, sdk.WithStack(fmt.Errorf("invalid vsphere hatchery configuration"))
+	datacenter, err := finder.DatacenterOrDefault(ctx, c.datacenter)
+	if err != nil {
+		return nil, sdk.WrapError(err, "unable to find datacenter %q", err)
 	}
+	finder.SetDatacenter(datacenter)
 
-	cfg.Host = sConfig.API.HTTP.URL
-	cfg.Token = sConfig.API.Token
-	cfg.InsecureSkipVerifyTLS = sConfig.API.HTTP.Insecure
-	cfg.RequestSecondsTimeout = sConfig.API.RequestTimeout
-	return cfg, nil
+	return finder, nil
 }
 
-func (h *HatcheryVSphere) GetLogger() *logrus.Logger {
-	return h.ServiceLogger
+func (c *vSphereClient) ListVirtualMachines(ctx context.Context) ([]mo.VirtualMachine, error) {
```

Comment:
I don't see any filter on CDS created ressource, for openstack we check if a label hatchery_name exists.

Reply:
this part of the code is a kind of low level vSphere client

Reply:
we are filtering CDS VM elsewhere with the function `getVirtualMachineCDSAnnotation`